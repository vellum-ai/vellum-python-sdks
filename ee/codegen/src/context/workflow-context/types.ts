export interface SDK_MODULE_PATHS {
  WORKFLOWS_MODULE_PATH: readonly string[];
  CORE_NODES_MODULE_PATH: readonly string[];
  DISPLAYABLE_NODES_MODULE_PATH: readonly string[];
  INPUTS_MODULE_PATH: readonly string[];
  SANDBOX_RUNNER_MODULE_PATH: readonly string[];
  STATE_MODULE_PATH: readonly string[];
  NODE_DISPLAY_MODULE_PATH: readonly string[];
  NODE_DISPLAY_TYPES_MODULE_PATH: readonly string[];
  VELLUM_TYPES_MODULE_PATH: readonly string[];
  PORTS_MODULE_PATH: readonly string[];
}
