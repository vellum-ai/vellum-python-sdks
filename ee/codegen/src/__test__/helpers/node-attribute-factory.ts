import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
