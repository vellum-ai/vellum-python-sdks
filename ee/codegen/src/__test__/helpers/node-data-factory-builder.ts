import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
