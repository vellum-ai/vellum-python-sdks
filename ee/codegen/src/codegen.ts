import {
  ErrorLogFile,
  InitFile,
  Inputs,
  NodeInput,
  VellumValue,
  VellumVariable,
  Workflow,
} from "./generators";
import { BasePersistedFile } from "./generators/base-persisted-file";
import { State } from "./generators/state";
import { WorkflowSandboxFile } from "./generators/workflow-sandbox-file";

export function vellumVariable(args: VellumVariable.Args): VellumVariable {
  return new VellumVariable(args);
}

export function vellumValue(args: VellumValue.Args): VellumValue {
  return new VellumValue(args);
}

export function nodeInput(args: NodeInput.Args): NodeInput {
  return new NodeInput(args);
}

export function inputs(args: Inputs.Args): Inputs {
  return new Inputs(args);
}

export function state(args: State.Args): State {
  return new State(args);
}

export function workflow(args: Workflow.Args): Workflow {
  return new Workflow(args);
}

export function initFile(args: InitFile.Args): InitFile {
  return new InitFile(args);
}

export function errorLogFile(args: BasePersistedFile.Args): ErrorLogFile {
  return new ErrorLogFile(args);
}

export function workflowSandboxFile(
  args: WorkflowSandboxFile.Args
): WorkflowSandboxFile {
  return new WorkflowSandboxFile(args);
}
