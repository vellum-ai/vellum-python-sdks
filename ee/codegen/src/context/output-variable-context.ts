import { VellumVariable } from "vellum-ai/api/types";

export declare namespace OutputVariableContext {
  export type Args = {
    outputVariableData: VellumVariable;
  };
}

export class OutputVariableContext {
  private readonly outputVariableData: VellumVariable;
  public readonly name: string;

  constructor({ outputVariableData }: OutputVariableContext.Args) {
    this.outputVariableData = outputVariableData;
    this.name = outputVariableData.key;
  }

  public getOutputVariableId(): string {
    return this.outputVariableData.id;
  }

  public getOutputVariableData(): VellumVariable {
    return this.outputVariableData;
  }

  public getRawName(): string {
    return this.outputVariableData.key;
  }
}
