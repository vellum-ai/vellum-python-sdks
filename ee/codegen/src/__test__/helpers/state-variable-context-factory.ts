import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
