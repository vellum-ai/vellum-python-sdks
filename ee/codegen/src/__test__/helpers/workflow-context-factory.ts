import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
