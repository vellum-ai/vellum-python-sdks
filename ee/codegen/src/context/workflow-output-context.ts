import { isEmpty, isNil } from "lodash";

import { WorkflowContext } from "src/context/workflow-context";
import { WorkflowOutputGenerationError } from "src/generators/errors";
import {
  FinalOutputNode as FinalOutputNodeType,
  WorkflowOutputValue as WorkflowOutputValueType,
} from "src/types/vellum";
import { toPythonSafeSnakeCase } from "src/utils/casing";

export declare namespace WorkflowOutputContext {
  export type Args = {
    workflowContext: WorkflowContext;
    terminalNodeData?: FinalOutputNodeType;
    workflowOutputValue?: WorkflowOutputValueType;
  };
}

export class WorkflowOutputContext {
  private readonly workflowContext: WorkflowContext;
  private readonly terminalNodeData?: FinalOutputNodeType;
  private readonly workflowOutputValue?: WorkflowOutputValueType;
  public readonly name: string;

  constructor({
    terminalNodeData,
    workflowContext,
    workflowOutputValue,
  }: WorkflowOutputContext.Args) {
    this.workflowContext = workflowContext;
    this.terminalNodeData = terminalNodeData;
    this.workflowOutputValue = workflowOutputValue;
    this.name = this.generateSanitizedOutputVariableName();
  }

  public getOutputNodeId(): string {
    let outputNodeId: string | undefined;
    if (this.workflowOutputValue) {
      outputNodeId = this.workflowOutputValue.value.nodeId;
    }
    if (this.terminalNodeData) {
      outputNodeId = this.terminalNodeData.id;
    }

    if (!outputNodeId) {
      throw new WorkflowOutputGenerationError(
        "Expected output node id in either output values or terminal data"
      );
    }
    return outputNodeId;
  }

  public getFinalOutputNodeData():
    | FinalOutputNodeType
    | WorkflowOutputValueType {
    if (this.workflowOutputValue) {
      return this.workflowOutputValue;
    } else if (this.terminalNodeData) {
      return this.terminalNodeData;
    } else {
      throw new WorkflowOutputGenerationError(
        "Expected either workflow output value or terminal node data to be defined"
      );
    }
  }

  private generateSanitizedOutputVariableName(): string {
    const defaultName = "output_";

    let rawOutputVariableName: string;
    if (this.workflowOutputValue) {
      const outputVariable = this.workflowContext.getOutputVariableContextById(
        this.workflowOutputValue.outputVariableId
      );
      rawOutputVariableName = outputVariable.name;
    } else if (this.terminalNodeData) {
      rawOutputVariableName = this.terminalNodeData.data.name;
    } else {
      throw new WorkflowOutputGenerationError(
        "Expected either workflow output value or terminal node data to be defined"
      );
    }

    const initialOutputVariableName =
      !isNil(rawOutputVariableName) && !isEmpty(rawOutputVariableName)
        ? toPythonSafeSnakeCase(rawOutputVariableName, "output")
        : defaultName;

    // Deduplicate the output variable name if it's already in use
    let sanitizedName = initialOutputVariableName;
    let numRenameAttempts = 0;
    while (this.workflowContext.isOutputVariableNameUsed(sanitizedName)) {
      sanitizedName = `${initialOutputVariableName}${
        initialOutputVariableName.endsWith("_") ? "" : "_"
      }${numRenameAttempts + 1}`;
      numRenameAttempts += 1;
    }

    return sanitizedName;
  }
}
