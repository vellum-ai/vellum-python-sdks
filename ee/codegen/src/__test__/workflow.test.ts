import { Writer } from "@fern-api/python-ast/core/Writer";
import { v4 as uuidv4 } from "uuid";
import { DocumentIndexRead } from "vellum-ai/api";
import { DocumentIndexes as DocumentIndexesClient } from "vellum-ai/api/resources/documentIndexes/client/Client";
import { afterEach, vi } from "vitest";

import { workflowContextFactory } from "./helpers";
import { edgesFactory } from "./helpers/edge-data-factories";
import { inputVariableContextFactory } from "./helpers/input-variable-context-factory";
import {
  conditionalNodeFactory,
  entrypointNodeDataFactory,
  finalOutputNodeFactory,
  searchNodeDataFactory,
  templatingNodeFactory,
  terminalNodeDataFactory,
} from "./helpers/node-data-factories";

import { mockDocumentIndexFactory } from "src/__test__/helpers/document-index-factory";
import { workflowOutputContextFactory } from "src/__test__/helpers/workflow-output-context-factory";
import * as codegen from "src/codegen";
import { createNodeContext, WorkflowContext } from "src/context";
import { OutputVariableContext } from "src/context/output-variable-context";
import { WorkflowOutputContext } from "src/context/workflow-output-context";
import { WorkflowGenerationError } from "src/generators/errors";
import { GraphAttribute } from "src/generators/graph-attribute";
import { WorkflowEdge } from "src/types/vellum";

describe("Workflow", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  const moduleName = "test";
  const entrypointNode = entrypointNodeDataFactory();

  beforeEach(async () => {
    vi.spyOn(DocumentIndexesClient.prototype, "retrieve").mockResolvedValue(
      mockDocumentIndexFactory() as unknown as DocumentIndexRead
    );

    workflowContext = workflowContextFactory();
    workflowContext.addEntrypointNode(entrypointNode);

    const nodeData = terminalNodeDataFactory();
    await createNodeContext({
      workflowContext,
      nodeData,
    });

    writer = new Writer();
  });

  afterEach(async () => {
    vi.restoreAllMocks();
  });

  describe("write", () => {
    it("should generate correct code when there are input variables", async () => {
      const inputs = codegen.inputs({ workflowContext });
      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code when there are no input variables", async () => {
      const inputs = codegen.inputs({ workflowContext });
      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code with Search Results as an output variable", async () => {
      workflowContext.addInputVariableContext(
        inputVariableContextFactory({
          inputVariableData: {
            id: "input-variable-id",
            key: "query",
            type: "STRING",
          },
          workflowContext,
        })
      );

      const inputs = codegen.inputs({ workflowContext });

      workflowContext.addOutputVariableContext(
        new OutputVariableContext({
          outputVariableData: {
            id: "output-id",
            key: "query",
            type: "STRING",
          },
          workflowContext,
        })
      );
      workflowContext.addWorkflowOutputContext(
        workflowOutputContextFactory({ workflowContext })
      );

      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should handle edges pointing to non-existent nodes", async () => {
      workflowContext.addInputVariableContext(
        inputVariableContextFactory({
          inputVariableData: {
            id: "input-variable-id",
            key: "query",
            type: "STRING",
          },
          workflowContext,
        })
      );

      const inputs = codegen.inputs({ workflowContext });

      const searchNodeData = searchNodeDataFactory();
      await createNodeContext({
        workflowContext: workflowContext,
        nodeData: searchNodeData,
      });

      const edges: WorkflowEdge[] = [
        {
          id: "edge-1",
          type: "DEFAULT",
          sourceNodeId: entrypointNode.id,
          sourceHandleId: entrypointNode.data.sourceHandleId,
          targetNodeId: searchNodeData.id,
          targetHandleId: searchNodeData.data.sourceHandleId,
        },
        {
          id: "edge-2",
          type: "DEFAULT",
          sourceNodeId: searchNodeData.id,
          sourceHandleId: "some-handle",
          targetNodeId: "non-existent-node-id",
          targetHandleId: "some-target-handle",
        },
      ];

      workflowContext.addWorkflowEdges(edges);

      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should handle the case of multiple nodes with the same label", async () => {
      const templatingNodeData1 = templatingNodeFactory({
        id: "7e09927b-6d6f-4829-92c9-54e66bdcaf80",
        label: "Templating Node",
        sourceHandleId: "dd8397b1-5a41-4fa0-8c24-e5dffee4fb98",
        targetHandleId: "3feb7e71-ec63-4d58-82ba-c3df829a2948",
      });
      await createNodeContext({
        workflowContext: workflowContext,
        nodeData: templatingNodeData1,
      });

      const templatingNodeData2 = templatingNodeFactory({
        id: "7e09927b-6d6f-4829-92c9-54e66bdcaf81",
        label: "Templating Node",
        sourceHandleId: "dd8397b1-5a41-4fa0-8c24-e5dffee4fb99",
        targetHandleId: "3feb7e71-ec63-4d58-82ba-c3df829a2949",
      });
      await createNodeContext({
        workflowContext: workflowContext,
        nodeData: templatingNodeData2,
      });

      const edges: WorkflowEdge[] = [
        {
          id: "edge-1",
          type: "DEFAULT",
          sourceNodeId: entrypointNode.id,
          sourceHandleId: entrypointNode.data.sourceHandleId,
          targetNodeId: templatingNodeData1.id,
          targetHandleId: templatingNodeData1.data.targetHandleId,
        },
        {
          id: "edge-2",
          type: "DEFAULT",
          sourceNodeId: templatingNodeData1.id,
          sourceHandleId: templatingNodeData1.data.sourceHandleId,
          targetNodeId: templatingNodeData2.id,
          targetHandleId: templatingNodeData2.data.targetHandleId,
        },
      ];
      workflowContext.addWorkflowEdges(edges);

      const inputs = codegen.inputs({ workflowContext });
      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    }, 1000000);

    it("should handle the workflow generation even if graph attribute fails in non-strict mode", async () => {
      vi.spyOn(
        GraphAttribute.prototype,
        "generateGraphMutableAst"
      ).mockImplementation(() => {
        throw new WorkflowGenerationError("test");
      });

      const workflowContext = workflowContextFactory({ strict: false });
      workflowContext.addEntrypointNode(entrypointNode);

      const templatingNodeData1 = templatingNodeFactory({
        id: "7e09927b-6d6f-4829-92c9-54e66bdcaf80",
        label: "Templating Node",
        sourceHandleId: "dd8397b1-5a41-4fa0-8c24-e5dffee4fb98",
        targetHandleId: "3feb7e71-ec63-4d58-82ba-c3df829a2948",
      });
      await createNodeContext({
        workflowContext: workflowContext,
        nodeData: templatingNodeData1,
      });

      const edges: WorkflowEdge[] = [
        {
          id: "edge-1",
          type: "DEFAULT",
          sourceNodeId: entrypointNode.id,
          sourceHandleId: entrypointNode.data.sourceHandleId,
          targetNodeId: templatingNodeData1.id,
          targetHandleId: templatingNodeData1.data.targetHandleId,
        },
      ];
      workflowContext.addWorkflowEdges(edges);

      const inputs = codegen.inputs({ workflowContext });
      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should handle loops of conditionals and import Graph.from_set properly", async () => {
      const startNode = conditionalNodeFactory({
        label: "Start Node",
      });
      await createNodeContext({
        workflowContext,
        nodeData: startNode,
      });

      const loopCheckNode = conditionalNodeFactory({
        id: uuidv4(),
        label: "Loop Check Node",
        ifSourceHandleId: uuidv4(),
        elseSourceHandleId: uuidv4(),
        targetHandleId: uuidv4(),
      });
      await createNodeContext({
        workflowContext,
        nodeData: loopCheckNode,
      });

      const finalOutput = finalOutputNodeFactory();
      await createNodeContext({
        workflowContext,
        nodeData: finalOutput,
      });

      const topNode = templatingNodeFactory({
        label: "Top Node",
      });
      await createNodeContext({
        workflowContext,
        nodeData: topNode,
      });

      workflowContext.addWorkflowEdges(
        edgesFactory([
          [entrypointNode, startNode],
          [[startNode, "0"], topNode],
          [[startNode, "1"], loopCheckNode],
          [[loopCheckNode, "0"], startNode],
          [[loopCheckNode, "1"], finalOutput],
          [[topNode, "0"], loopCheckNode],
        ])
      );

      const inputs = codegen.inputs({ workflowContext });
      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code with an output variable mapped to an input variable", async () => {
      workflowContext.addInputVariableContext(
        inputVariableContextFactory({
          inputVariableData: {
            id: "input-variable-id",
            key: "query",
            type: "STRING",
          },
          workflowContext,
        })
      );

      const inputs = codegen.inputs({ workflowContext });

      workflowContext.addOutputVariableContext(
        new OutputVariableContext({
          outputVariableData: {
            id: "output-variable-id",
            key: "passthrough",
            type: "STRING",
          },
          workflowContext,
        })
      );

      workflowContext.addWorkflowOutputContext(
        new WorkflowOutputContext({
          workflowOutputValue: {
            outputVariableId: "output-variable-id",
            value: {
              type: "WORKFLOW_INPUT",
              inputVariableId: "input-variable-id",
            },
          },
          workflowContext,
        })
      );

      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct display code when there are input variables with escape characters", async () => {
      const entrypointNode = entrypointNodeDataFactory();
      workflowContext = workflowContextFactory();
      workflowContext.addEntrypointNode(entrypointNode);

      const nodeData = terminalNodeDataFactory();
      await createNodeContext({
        workflowContext,
        nodeData,
      });
      workflowContext.addInputVariableContext(
        inputVariableContextFactory({
          inputVariableData: {
            id: "5f64210f-ec43-48ce-ae40-40a9ba4c4c11",
            key: 'Bad \\"Input\\"',
            type: "STRING",
          },
          workflowContext,
        })
      );
      const inputs = codegen.inputs({ workflowContext });
      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code with an output variable mapped to a non-existent node", async () => {
      const workflowContext = workflowContextFactory({ strict: false });
      workflowContext.addInputVariableContext(
        inputVariableContextFactory({
          inputVariableData: {
            id: "input-variable-id",
            key: "query",
            type: "STRING",
          },
          workflowContext,
        })
      );

      const inputs = codegen.inputs({ workflowContext });

      workflowContext.addOutputVariableContext(
        new OutputVariableContext({
          outputVariableData: {
            id: "output-variable-id",
            key: "passthrough",
            type: "STRING",
          },
          workflowContext,
        })
      );

      workflowContext.addWorkflowOutputContext(
        new WorkflowOutputContext({
          workflowOutputValue: {
            outputVariableId: "output-variable-id",
            value: {
              type: "NODE_OUTPUT",
              nodeId: "non-existent-node-id",
              nodeOutputId: "output-id",
            },
          },
          workflowContext,
        })
      );

      const workflow = codegen.workflow({
        moduleName,
        workflowContext,
        inputs,
      });

      workflow.getWorkflowFile().write(writer);
      expect(await writer.toString()).toMatchSnapshot();
      const errors = workflowContext.getErrors();

      expect(errors.length).toEqual(1);
      expect(errors[0]?.message).toEqual(
        "Failed to find node with id 'non-existent-node-id'"
      );
      expect(errors[0]?.severity).toEqual("WARNING");
    });
  });
});
