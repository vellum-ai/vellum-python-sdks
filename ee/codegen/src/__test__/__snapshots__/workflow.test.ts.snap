// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Workflow > write > should generate correct code when there are input variables 1`] = `
"from vellum.workflows import BaseWorkflow
from vellum.workflows.state import BaseState

from .inputs import Inputs


class TestWorkflow(BaseWorkflow[Inputs, BaseState]):
    pass
"
`;

exports[`Workflow > write > should generate correct code when there are no input variables 1`] = `
"from vellum.workflows import BaseWorkflow


class TestWorkflow(BaseWorkflow):
    pass
"
`;

exports[`Workflow > write > should generate correct code with Search Results as an output variable 1`] = `
"from vellum.workflows import BaseWorkflow
from vellum.workflows.state import BaseState

from .inputs import Inputs
from .nodes.final_output import FinalOutput


class TestWorkflow(BaseWorkflow[Inputs, BaseState]):
    class Outputs(BaseWorkflow.Outputs):
        query = FinalOutput.Outputs.value
"
`;

exports[`Workflow > write > should generate correct code with an output variable mapped to a non-existent node 1`] = `
"from vellum.workflows import BaseWorkflow
from vellum.workflows.state import BaseState

from .inputs import Inputs


class TestWorkflow(BaseWorkflow[Inputs, BaseState]):
    class Outputs(BaseWorkflow.Outputs):
        passthrough = None
"
`;

exports[`Workflow > write > should generate correct code with an output variable mapped to an input variable 1`] = `
"from vellum.workflows import BaseWorkflow
from vellum.workflows.state import BaseState

from .inputs import Inputs


class TestWorkflow(BaseWorkflow[Inputs, BaseState]):
    class Outputs(BaseWorkflow.Outputs):
        passthrough = Inputs.query
"
`;

exports[`Workflow > write > should generate correct code with an output variable mapped to an unused terminal node 1`] = `
"from vellum.workflows import BaseWorkflow

from .nodes.final_output import FinalOutput


class TestWorkflow(BaseWorkflow):
    unused_graphs = {FinalOutput}

    class Outputs(BaseWorkflow.Outputs):
        passthrough = FinalOutput.Outputs.value
"
`;

exports[`Workflow > write > should generate correct display code when there are input variables with escape characters 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.base import (
    WorkflowDisplayData,
    WorkflowDisplayDataViewport,
    WorkflowInputsDisplay,
    WorkflowMetaDisplay,
)
from vellum_ee.workflows.display.editor import NodeDisplayData, NodeDisplayPosition
from vellum_ee.workflows.display.workflows import BaseWorkflowDisplay

from ..inputs import Inputs
from ..workflow import TestWorkflow


class TestWorkflowDisplay(BaseWorkflowDisplay[TestWorkflow]):
    workflow_display = WorkflowMetaDisplay(
        entrypoint_node_id=UUID("entrypoint"),
        entrypoint_node_source_handle_id=UUID("<source_handle_id>"),
        entrypoint_node_display=NodeDisplayData(
            position=NodeDisplayPosition(x=0, y=0),
            z_index=None,
            width=None,
            height=None,
        ),
        display_data=WorkflowDisplayData(
            viewport=WorkflowDisplayDataViewport(x=0, y=0, zoom=0)
        ),
    )
    inputs_display = {
        Inputs.bad_input: WorkflowInputsDisplay(
            id=UUID("5f64210f-ec43-48ce-ae40-40a9ba4c4c11"), name='Bad "Input"'
        )
    }
    entrypoint_displays = {}
    output_displays = {}
"
`;

exports[`Workflow > write > should handle edges pointing to non-existent nodes 1`] = `
"from vellum.workflows import BaseWorkflow
from vellum.workflows.state import BaseState

from .inputs import Inputs
from .nodes.search_node import SearchNode


class TestWorkflow(BaseWorkflow[Inputs, BaseState]):
    graph = SearchNode
"
`;

exports[`Workflow > write > should handle loops of conditionals and import Graph.from_set properly 1`] = `
"from vellum.workflows import BaseWorkflow
from vellum.workflows.graph import Graph

from .nodes.final_output_node import FinalOutputNode
from .nodes.loop_check_node import LoopCheckNode
from .nodes.start_node import StartNode
from .nodes.top_node import TopNode


class TestWorkflow(BaseWorkflow):
    graph = {
        StartNode.Ports.if_1 >> TopNode,
        StartNode.Ports.else_1,
    } >> Graph.from_set(
        {
            LoopCheckNode.Ports.if_1 >> StartNode,
            LoopCheckNode.Ports.else_1 >> FinalOutputNode,
        }
    )
"
`;

exports[`Workflow > write > should handle the case of multiple nodes with the same label 1`] = `
"from vellum.workflows import BaseWorkflow

from .nodes.templating_node import TemplatingNode
from .nodes.templating_node_1 import TemplatingNode1


class TestWorkflow(BaseWorkflow):
    graph = TemplatingNode >> TemplatingNode1
"
`;

exports[`Workflow > write > should handle the workflow generation even if graph attribute fails in non-strict mode 1`] = `
"from vellum.workflows import BaseWorkflow


class TestWorkflow(BaseWorkflow):
    pass
"
`;
